### DISCUSSION TITLE
Highlighting of custom treesitter queries for markdown does not work as expected.

### DISCUSSION THREAD

**Discussion by camoz** [neovim/neovim#29070] [Q&A] [answered]
Hi,

I have a problem with the highlighting of custom queries. I have a custom queries for markdown header markers in `.config/nvim/after/queries/markdown/highlights.scm`: 

```
;; extends

(atx_heading (atx_h1_marker) @markup.heading.1.marker)
(atx_heading (atx_h2_marker) @markup.heading.2.marker)
(atx_heading (atx_h3_marker) @markup.heading.3.marker)
(atx_heading (atx_h4_marker) @markup.heading.4.marker)
(atx_heading (atx_h5_marker) @markup.heading.5.marker)
(atx_heading (atx_h6_marker) @markup.heading.6.marker)
```

In my colorscheme, I have set `gui=bold,reverse` for the level 1 *headers* `@markup.heading.1.markdown` but I have `gui=none` for the level 1 *header markers* `@markup.heading.1.marker.markdown`.

In the output of `:filter markup.heading highlight` it is shown correctly, but in the buffer it is not highlighted correctly: The level 1 header marker is `bold` and `reverse`, and the level 2 header marker is `bold`, contrary to expectation. The `gui=` values of the *headers* seem to get applied to the *header markers*, in my case it is visible for level 1 and level 2.

![screen_2024-05-28_120129](https://github.com/neovim/neovim/assets/6276613/21e558f7-e2df-46f1-a5c3-f8936a0134a2)

Any idea why this is the case?

**Comment by clason** [comment]
Don't put them in `after`; the `;extends` line is enough. (This was changed _a long_ time ago.)

**Comment by camoz** [comment]
I've moved it out of the `after` folder into `.config/nvim/queries/markdown/highlights.scm`, but sadly that doesn't change anything... thank you for the clarification though!

**Comment by clason** [comment]
What does `:Inspect` show on the different markers?

**Comment by camoz** [comment]
![screen_2024-05-28_211801](https://github.com/neovim/neovim/assets/6276613/b843373c-ace6-4940-9bff-3e25e085a343)

The `@markup.heading.*.marker.markdown` is always the same for every level like in the screenshot above, and the actual headings `@markup.heading.*.markdown` are highlighted as in the buffer.

Edit: I made this screenshot bigger so that one can see that the level 2 heading marker is displayed bold.

**Reply by clason** [reply]
That means your query is not applied. This could be an issue with non-standard runtime paths.

**Reply by camoz** [reply]
Hmm I don't follow. If I omit my queries/captures, the header markers are displayed quite differently, namely like the headings... so I'm sure my queries get applied?

**Comment by camoz** [comment]
I wonder if this is intended behavior, because the highlights of `@markup.heading.*.marker.markdown` have no explicit `gui=` set (see screenshot in OP) and thus inherit it from the `gui=` of `@markup.heading.*.markdown`? If that makes sense? I've tried `:hi @markup.heading.1.marker.markdown gui=NONE` but the `gui=NONE` doesn't seem to have an effect (it does not get added to the output of `:highlight`).

**Comment by camoz** [comment]
Ok, so if I add my captures for the heading markers to the `highlights.scm` shipped with nvim-treesitter, it gets highlighted correctly! I guess the problem is that the markers get captured as `@markup.heading.1` by nvim-treesitter?

Excerpt:

```
(atx_heading
  (atx_h1_marker) @markup.heading.1
  (inline) @markup.heading.1)
```

**Comment by camoz** [comment]
Hmm I don't follow. If I omit my queries/captures, the header markers are displayed quite differently, namely like the headings... so I'm sure my queries get applied?

**Comment by clason** [comment]
How do you define your highlight groups then?

**Reply by camoz** [reply]
Via the onedark colorscheme setup function, but it's the same if I delete them there and use `:hi @markup.heading.3.marker.markdown guifg=grey gui=NONE` etc. instead.

(Sorry for double post, was replying via e-mail earlier and it didn't move my reply to your post)

**Comment by clason** [comment, marked as answer]
The point here is that if you extend, you apply your capture on top of the previous capture. And depending on how you define the highlight groups, you will get a combination of attributes.

**Reply by camoz** [reply]
I see, so I guess there is no way to remove the `gui=` formatting that I do for `@markup.heading.*.markdown` for any `@markup.heading.*.marker.markdown` that I define afterwards?

**Reply by clason** [reply]
You can define the full set of attributes for the highlight group. (Or make sure combine isn't used.)

I can't help you with colorscheme plugins; you'll have to raise the issue with them.

**Reply by camoz** [reply]
Ofc I understand that you cannot help with plugins, that's why I wrote it's the same when doing the highlight manually via `:highlight`.

As I said, it does not work if I do e.g. `:hi @markup.heading.3.marker.markdown guifg=grey gui=NONE`, the `gui=NONE` seems to be ignored... so the "full set of attributes for the highlight group" does not seem to take effect somehow.

Btw I appreciate your help! I will do more testing tomorrow, have to go afk now... but thank you so far :)

**Reply by clason** [reply]
As I said, both highlight groups are applied in order. So you need to make sure that your highlight group doesn't combine with the previous one `:h nocombine`.

**Reply by camoz** [reply]
Thank you for the pointer to `:h nocombine`, I didn't completely understand what you meant with "make sure combine isn't used" earlier, I sould have read more carefully. I didn't know this was a thing and it all makes sense now! Thank you!
